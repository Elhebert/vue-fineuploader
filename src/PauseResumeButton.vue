<template>
  <renderless-pause-resume-button
    v-bind="{id, uploader, allowPauseOnlyAfterFirstChunk}"
  >
    <button
      v-if="pausable || resumable || !onlyRenderIfEnabled"
      slot-scope="{ pausable, resumable, buttonEvents }"
      :disabled="!pausable && !resumable"
      v-on="buttonEvents"
    >
      <template v-if="resumable">Resume</template>
      <template v-else>Pause</template>
    </button>
  </renderless-pause-resume-button>
</template>

<script>
import RenderlessPauseResumeButton from './renderless/PauseResumeButton'

export default {
  components: {
    RenderlessPauseResumeButton,
  },

  props: {
    id: {
      type: Number,
      required: true,
    },
    onlyRenderIfEnabled: {
      type: Boolean,
      default: true,
    },
    allowPauseOnlyAfterFirstChunk: {
      type: Boolean,
      default: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },
}
</script>
