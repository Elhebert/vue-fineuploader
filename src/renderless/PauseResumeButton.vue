<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    allowPauseOnlyAfterFirstChunk: {
      type: Boolean,
      default: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  data: () => ({
    atLeastOneChunkUploaded: false,
    pausable: false,
    resumable: false,
    unmounted: false,
    statuses: null,
    initialStatus: null,
  }),

  computed: {
    buttonClassName() {
      return this.resumable
        ? 'fine-uploader-resume-button'
        : 'fine-uploader-pause-button'
    },

    buttonLabel() {
      return this.resumable ? 'resume' : 'pause'
    },
  },

  created() {
    this.statuses = this.uploader.qq.status
    this.initialStatus = this.uploader.methods.getUploads({
      id: this.id,
    }).status
    this.pausable = this.initialStatus === this.statuses.UPLOADING
    this.resumable = this.initialStatus === this.statuses.PAUSED
  },

  mounted() {
    this.uploader.on('resume', this.onResume)
    this.uploader.on('statusChange', this.onStatusChange)
    this.uploader.on('uploadChunk', this.onUploadChunk)
  },

  beforeDestroy() {
    this.unmounted = true
    this.unregisterOnResumeHandler()
    this.unregisterOnUploadChunkHandler()
    this.unregisterOnStatusChangeHandler()
  },

  methods: {
    isPausable(status) {
      return status === this.statuses.UPLOADING
    },

    isResumable(status) {
      return status === this.statuses.PAUSED
    },

    onStatusChange(id, oldStatus, newStatus) {
      if (id !== this.id || this.unmounted) {
        return
      }
      const pausable = this.isPausable(newStatus)
      const resumable = !pausable && this.isResumable(newStatus)

      this.pausable = pausable
      this.resumable = resumable

      if (
        newStatus === this.statuses.DELETED ||
        newStatus === this.statuses.CANCELED ||
        newStatus === this.statuses.UPLOAD_SUCCESSFUL
      ) {
        this.unregisterOnResumeHandler()
        this.unregisterOnUploadChunkHandler()
        this.unregisterOnStatusChangeHandler()
      }
    },

    onClick(event) {
      event.preventDefault()

      if (this.pausable) {
        this.uploader.methods.pauseUpload(this.id)
      } else if (this.resumable) {
        this.uploader.methods.continueUpload(this.id)
      }
    },

    onUploadChunk(id, name, chunkData) {
      if (
        id === this.id &&
        this.allowPauseOnlyAfterFirstChunk &&
        !this.unmounted
      ) {
        if (chunkData.partIndex > 0 && !this.pausable) {
          this.pausable = true
          this.resumable = false
        } else if (chunkData.partIndex === 0 && this.pausable) {
          this.pausable = false
          this.resumable = false
        }
      }
    },

    onResume(id) {
      if (id === this.id && !this.unmounted && !this.atLeastOneChunkUploaded) {
        this.atLeastOneChunkUploaded = true
        this.pausable = true
        this.resumable = false
      }
    },

    unregisterOnResumeHandler() {
      this.uploader.off('resume', this.onResume)
    },

    unregisterOnStatusChangeHandler() {
      this.uploader.off('statusChange', this.onStatusChange)
    },

    unregisterOnUploadChunkHandler() {
      this.uploader.off('uploadChunk', this.onUploadChunk)
    },
  },

  render() {
    return this.$scopedSlots.default({
      pausable: this.pausable,
      resumable: this.resumable,
      buttonEvents: {
        click: event => {
          this.onClick(event)
        },
      },
    })
  },
}
</script>
