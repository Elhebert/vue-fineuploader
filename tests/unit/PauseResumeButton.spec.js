import { mount } from '@vue/test-utils'
import PauseResumeButton from '@/PauseResumeButton.vue'
import FineUploader from 'fine-uploader-wrappers'
import { status as STATUSES } from 'fine-uploader/lib/core/all'

describe('PauseResumeButton.vue', () => {
  let wrapper
  let renderless
  let uploader

  beforeEach(() => {
    uploader = new FineUploader({ options: {} })

    wrapper = mount(PauseResumeButton, {
      propsData: { id: 0, uploader },
    })

    renderless = wrapper.vm.$children[0]
  })

  it("doesn't render the button until the first chunk has been uploaded", () => {
    expect(wrapper.isEmpty()).toBe(true)
    expect(wrapper).toMatchSnapshot()

    renderless.onUploadChunk(0, null, { partIndex: 3 })

    expect(wrapper.html()).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('removes the pause button when the upload is no longer actionable', () => {
    renderless.onUploadChunk(0, null, { partIndex: 1 })

    expect(wrapper.html()).toBeDefined()
    expect(wrapper).toMatchSnapshot()

    renderless.onStatusChange(0, null, STATUSES.DELETED)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toBeUndefined()
  })

  it('disables the button, if requested, when the upload is no longer actionable', () => {
    wrapper.setProps({ onlyRenderIfEnabled: false })

    renderless.onUploadChunk(0, null, { partIndex: 1 })

    expect(wrapper.html()).toBeDefined()
    expect(wrapper).toMatchSnapshot()

    renderless.onStatusChange(0, null, STATUSES.DELETED)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toBeDefined()
    expect(wrapper.attributes().disabled).toBeDefined()
  })
})
