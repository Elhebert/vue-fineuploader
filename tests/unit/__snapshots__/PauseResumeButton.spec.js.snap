// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PauseResumeButton.vue disables the button, if requested, when the upload is no longer actionable 1`] = `<button>Pause</button>`;

exports[`PauseResumeButton.vue disables the button, if requested, when the upload is no longer actionable 2`] = `<button disabled="disabled">Pause</button>`;

exports[`PauseResumeButton.vue doesn't render the button until the first chunk has been uploaded 1`] = ``;

exports[`PauseResumeButton.vue doesn't render the button until the first chunk has been uploaded 2`] = `<button>Pause</button>`;

exports[`PauseResumeButton.vue removes the pause button when the upload is no longer actionable 1`] = `<button>Pause</button>`;

exports[`PauseResumeButton.vue removes the pause button when the upload is no longer actionable 2`] = ``;
